### ISSUE TITLE
MooseUtils::pathjoin needs to be improved

### ISSUE THREAD

**Issue by YaqiWang** [idaholab/moose#21150]
## Reason
I am seeing something like `moose/framework/data/../../share/moose/framework/data/../../../foo` during accessing a file.

## Design
We can simply improve `MooseUtils::pathjoin` a little to get rid of this and make the screen output sane. The improvement should also protect illegal access like `moose/framework/../../../../foo`.

## Impact
Better code and should not impact current capabilities.

**Comment by dschwen** [comment]
We can mess around with this, but the correct solution is in https://github.com/idaholab/moose/pull/21137 . `std::filesystem::path` can perform this sort of normalization.

**Comment by YaqiWang** [comment]
How likely that PR can be merged? Otherwise, I can push up a quick fix. It does not cost me much.

**Comment by dschwen** [comment]
yeah, you should go with the fix. std::filesystem is still a bit out, as it required more modern minimal compiler targets.
